app/helpers/es.py

```diff
@@ -636,3 +667,62 @@ def build_search_query(bool_clause=None,
         query["highlight"] = highlight_settings
 
     return query
+
+
+def build_first_occur_search_query(search_query, search_range, target_list, aggregator_list, timestamp):
+    """
+    Build specific Elasticsearch query for retrieving aggregation of first occurrence of a term field in events matching
+    with the search_query.
+
+    :param search_query: search query
+    :param search_range: search range
+    :param target_list: list of target fields
+    :param aggregator_list: list of fields that we want to aggregate
+    :param timestamp: timestamp field name
+    :return: the build Elasticsearch query
+    """
+
+    order_dict = dict()
+    order_dict[timestamp] = dict()
+    order_dict[timestamp]["order"] = "asc"
+
+    query = dict()
+
+    query["size"] = 0
+
+    query["query"] = dict()
+    query["query"]["bool"] = dict()
+    query["query"]["bool"]["filter"] = list()
+    query["query"]["bool"]["filter"].append(search_range)
+    query["query"]["bool"]["filter"].append(search_query["filter"])
+
+    query["aggs"] = dict()
+    query["aggs"]["aggregator"] = dict()
+    query["aggs"]["aggregator"]["terms"] = dict()
+    query["aggs"]["aggregator"]["terms"]["script"] = build_script(aggregator_list)
+
+    query["aggs"]["aggregator"]["aggs"] = dict()
+    query["aggs"]["aggregator"]["aggs"]["target"] = dict()
+    query["aggs"]["aggregator"]["aggs"]["target"]["terms"] = dict()
+    query["aggs"]["aggregator"]["aggs"]["target"]["terms"]["script"] = build_script(target_list)
```

Comment:
You can use default dict to auto initialize nested dicts: https://stackoverflow.com/questions/651794/whats-the-best-way-to-initialize-a-dict-of-dicts-in-python

https://docs.python.org/2/library/collections.html

Carefully read on the details on how they behave compared to regular dicts!

Reply:
As an alternative, can I define the dictionary as follow?
It will be more clear for future developer to compare it directly with DSL queries.
```javascript
     query = {"size": 0,
              "query": {
                  "bool": {
                      "filter": filter_list
                  }
              },
              "aggs": {
                  "aggregator": {
                      "terms": {
                          "script": build_script(aggregator_list)
                      },
                      "aggs": {
                          "target": {
                              "terms": {
                                  "script":  build_script(target_list)
                              },
                              "aggs": {
                                  "top_doc": {
                                      "top_hits": {
                                          "sort": [order_dict],
                                          "size": 1
                                      }
                                  }
                              }
                          }
                      }
                  }
              }}
```

Reply:
Sounds good!